ompi/mca/coll/xhc/coll_xhc.c

```diff
@@ -0,0 +1,700 @@
+    peer_cico_ds = malloc(comm_size * sizeof(opal_shmem_ds_t));
+    if(!data || !peer_cico_ds) {
+        RETURN_WITH_ERROR(return_code, OMPI_ERR_OUT_OF_RESOURCE, end);
+    }
+
+    *data = (xhc_data_t) {
+        .comms = NULL,
+        .comm_count = -1,
+
+        .pvt_coll_seq = 0
+    };
+
+    // ----
+
+    if(OMPI_XHC_CICO_MAX > 0) {
+        opal_shmem_ds_t cico_ds;
+
+        void *my_cico = xhc_shmem_create(&cico_ds,
+            OMPI_XHC_CICO_MAX, comm, "cico", 0);
+        if(!my_cico) {
+            RETURN_WITH_ERROR(return_code, OMPI_ERR_OUT_OF_RESOURCE, end);
+        }
+
+        /* Manually "touch" to assert allocation in local NUMA node
+         * (assuming linux's default firt-touch-alloc NUMA policy) */
+        memset(my_cico, 0, OMPI_XHC_CICO_MAX);
+
+        ret = comm->c_coll->coll_allgather(&cico_ds,
+            sizeof(opal_shmem_ds_t), MPI_BYTE, peer_cico_ds,
+            sizeof(opal_shmem_ds_t), MPI_BYTE, comm,
+            comm->c_coll->coll_allgather_module);
+        if(ret != OMPI_SUCCESS) {
+            RETURN_WITH_ERROR(return_code, ret, end);
+        }
+
+        for(int r = 0; r < comm_size; r++) {
+            peer_info[r].cico_ds = peer_cico_ds[r];
+        }
+
+        peer_info[rank].cico_buffer = my_cico;
+    }
+
+    // ----
+
+    /* An XHC communicator is created for each level of the hierarchy.
+     * The hierachy must be in an order of most-specific to most-general. */
+
+    ret = xhc_make_comms(comm, peer_info, &data->comms, &data->comm_count,
+        module->hierarchy, module->hierarchy_len);
+    if(ret != OMPI_SUCCESS) {
+        RETURN_WITH_ERROR(return_code, ret, end);
+    }
+
+    for(int i = 0, c = 0; i < data->comm_count; i++) {
+        data->comms[i].chunk_size = module->chunks[c];
+        c = opal_min(c + 1, module->chunks_len - 1);
+    }
+
+    if(module->chunks_len < data->comm_count) {
+        opal_output_verbose(MCA_BASE_VERBOSE_WARN,
+            ompi_coll_base_framework.framework_output,
+            "coll:xhc: Warning: The chunk sizes count is shorter than the "
+            "hierarchy size; filling in with the last entry provided");
+    } else if(module->chunks_len > data->comm_count) {
+        opal_output_verbose(MCA_BASE_VERBOSE_WARN,
+            ompi_coll_base_framework.framework_output,
+            "coll:xhc: Warning: The chunk size count is larger than the "
+            "hierarchy size; omitting last entries");
+    }
+
+    // ----
+
+    if(mca_coll_xhc_component.print_info) {
+        xhc_print_info(module, comm, data);
+    }
+
+    // ----
+
+    module->data = data;
+    module->init = true;
+
+    end:
+
+    xhc_module_set_coll_fns(comm, xhc_fns);
+
+    free(peer_cico_ds);
+
+    if(return_code != 0) {
+        opal_show_help("help-coll-xhc.txt", "xhc-init-failed", true,
+            return_code, errno, strerror(errno));
+
+        xhc_deinit(module);
+    }
+
+    return return_code;
+}
+
+void xhc_deinit(mca_coll_xhc_module_t *module) {
```

Comment:
OK, I will adjust it. There are also a number of other functions that are not static and begin with `xhc_`, so I'll adjust those as well.

Reply:
I renamed it to `xhc_fini` to match the other coll components. I also fixed the naming of all non-static functions.

For the functions that are non-static because they are called across xhc files, but are only intended to be used internally, I added some macros that omit the `mca_coll_` prefix, so as to keeps the names short and wieldy, and to offer a slight hint to their internal nature.